### PR TITLE
Range Infrastructure

### PR DESCRIPTION
**Pull request by parsa**
This PR adds support for ranges. Ranges can be an integer range or a range of arguments.

### CODE REVIEW

File: src/ir/ranges.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            return *(util::get<1>(it_));
+        }
+
+        HPX_THROW_EXCEPTION(hpx::invalid_status,
+            "phylanx::ir::range_iterator::dereference",
+            "range_iterator object holds unsupported data type");
+    }
+
+    bool range_iterator::equal(range_iterator const& other) const
+    {
```
Comment (hkaiser) [inline src/ir/ranges.cpp]: Please make sure both iterators hold the same type (index) before comparing the values themselves.

File: src/ir/ranges.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        case 0:    // int_range_type
+        {
+            int_range_type& p = util::get<0>(it_);
+            p.first += p.second;
+            break;
+        }
+        case 1:    // arg_range_type
+            ++util::get<1>(it_);
+            break;
+        default:
```
Comment (hkaiser) [inline src/ir/ranges.cpp]: This misses the case for `index() == 2`.

File: phylanx/ir/ranges.hpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            execution_tree::primitive_argument_type>
+    {
+    private:
+        using int_range_type = std::pair<std::int64_t, std::int64_t>;
+        using arg_range_type =
+            std::vector<execution_tree::primitive_argument_type>::iterator;
+        using iterator_type = util::variant<int_range_type, arg_range_type>;
+
+    public:
+        range_iterator(int_range_type it)
```
Comment (hkaiser) [inline phylanx/ir/ranges.hpp]: This constructor is not used anywhere, do we need it?

File: tests/unit/ir/range_iterator.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    std::vector<arg_t> v{
+        arg_t{static_cast<std::int64_t>(6)},
+        arg_t{static_cast<std::int64_t>(9)},
+        arg_t{static_cast<std::int64_t>(42)}};
+
+    phylanx::ir::range_iterator it(v.begin());
+    HPX_TEST_EQ(*it++, 6);
+    HPX_TEST_EQ(*it++, 9);
+    HPX_TEST_EQ(*it++, 42);
+}
```
Comment (hkaiser) [inline tests/unit/ir/range_iterator.cpp]: Please verify that `it == range_iterator{v.end()}`

File: src/ir/ranges.cpp (hunk 1 above)
Comment (parsa) [reply src/ir/ranges.cpp]: Done.

File: phylanx/ir/ranges.hpp (hunk 3 above)
Comment (parsa) [reply phylanx/ir/ranges.hpp]: Removed.

File: tests/unit/ir/range_iterator.cpp (hunk 4 above)
Comment (parsa) [reply tests/unit/ir/range_iterator.cpp]: Done.

File: phylanx/ir/ranges.hpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+        range_iterator begin();
+        range_iterator end();
+
+        reverse_range_iterator rbegin();
+        reverse_range_iterator rend();
+
+        explicit range(
+            std::vector<execution_tree::primitive_argument_type> data)
+          : data_(data)
```
Comment (hkaiser) [inline phylanx/ir/ranges.hpp]: This should be `std::move(data)`. Alternatively you might want to provide two constructors, one for `std::vector<execution_tree::primitive_argument_type> const& data` and one for `std::vector<execution_tree::primitive_argument_type> && data`.

File: src/ir/ranges.cpp [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+            std::int64_t start = hpx::util::get<0>(int_range);
+            std::int64_t step = hpx::util::get<2>(int_range);
+            return range_iterator{start, step};
+        }
+        case 1:    // args_type
+            return util::get<1>(data_).begin();
+        case 2:    // arg_pair_type
+            return util::get<2>(data_).first;
+        case 3:
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
```
Comment (hkaiser) [inline src/ir/ranges.cpp]: Hmmm. A range should not know whether it is a 'reverse' range or not. Also ranges are always 'forward' ranges. This makes it always be possible to derive a reverse iterator from a forward range.

File: phylanx/ir/ranges.hpp (hunk 5 above)
Comment (parsa) [reply phylanx/ir/ranges.hpp]: Done.

File: src/ir/ranges.cpp (hunk 6 above)
Comment (parsa) [reply src/ir/ranges.cpp]: I think I addressed it. We can't do `range x(v.rbegin(), v.rend()` anymore, though.

**Pull request merged**